Core/src/org/sleuthkit/autopsy/directorytree/DataResultFilterNode.java

```diff
@@ -240,22 +241,107 @@ private DataResultFilterChildren(Node arg, ExplorerManager sourceEm) {
         protected Node[] createNodes(Node key) {
             // if displaying the results from the Data Source tree
             // filter out artifacts
-          
+
             // In older versions of Autopsy,  attachments were children of email/message artifacts
             // and hence email/messages with attachments are shown in the tree data source tree,
             BlackboardArtifact art = key.getLookup().lookup(BlackboardArtifact.class);
             if (art != null && filterArtifacts
                     && ((FilterNodeUtils.showMessagesInDatasourceTree() == false)
-                         || (FilterNodeUtils.showMessagesInDatasourceTree()
-                                && art.getArtifactTypeID() != BlackboardArtifact.ARTIFACT_TYPE.TSK_EMAIL_MSG.getTypeID()
-                                && art.getArtifactTypeID() != BlackboardArtifact.ARTIFACT_TYPE.TSK_MESSAGE.getTypeID()))) {
+                    || (FilterNodeUtils.showMessagesInDatasourceTree()
+                    && art.getArtifactTypeID() != BlackboardArtifact.ARTIFACT_TYPE.TSK_EMAIL_MSG.getTypeID()
+                    && art.getArtifactTypeID() != BlackboardArtifact.ARTIFACT_TYPE.TSK_MESSAGE.getTypeID()))) {
                 return new Node[]{};
             }
-                
+
             return new Node[]{new DataResultFilterNode(key, sourceEm)};
         }
     }
 
+    /**
+     * An action that navigates to an artifact.
+     */
+    private static class ViewArtifactAction extends AbstractAction {
+
+        private final BlackboardArtifact artifact;
+
+        /**
+         * Main constructor.
+         *
+         * @param artifact    The artifact to navigate to in the action.
+         * @param displayName The display name of the menu item.
+         */
+        ViewArtifactAction(BlackboardArtifact artifact, String displayName) {
+            super(displayName);
+            this.artifact = artifact;
+        }
+
+        @Override
+        public void actionPerformed(ActionEvent e) {
+            DirectoryTreeTopComponent.findInstance().viewArtifact(artifact);
+        }
+    }
+
+    /**
+     * An action that navigates to an os account.
+     */
+    private static class ViewOsAccountAction extends AbstractAction {
+
+        private final OsAccount osAccount;
+
+        /**
+         * Main constructor.
+         *
+         * @param osAccount   The os account to navigate to in the action.
+         * @param displayName The display name of the menu item.
+         */
+        ViewOsAccountAction(OsAccount osAccount, String displayName) {
+            super(displayName);
+            this.osAccount = osAccount;
+        }
+
+        @Override
+        public void actionPerformed(ActionEvent e) {
+            DirectoryTreeTopComponent.findInstance().viewOsAccount(osAccount);
```

Comment:
This call should not be happening on the EDT! Call in SwingWorker, make sure that cursors appear to block any further tree actions.